processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java

```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
+
+    @Override
+    public Iterator<Range> iterator()
+    {
+      return new Iterator<Range>()
+      {
+        int currentRowIndex = 0;
+        int currentRangeIndex = 0;
+
+        @Override
+        public boolean hasNext()
+        {
+          return currentRowIndex < numRows;
+        }
+
+        @Override
+        public Range next()
+        {
+          if (!hasNext()) {
+            throw new IllegalStateException();
+          }
+          // TODO: invert listing order at the end to get benefits of incremenental aggregations
+          Range r = new Range(
+              Interval.of(
+                  rangeToRowIndex(relativeRangeId(0)),
+                  rangeToRowIndex(relativeRangeId(1))
+              ),
+              Interval.of(
+                  rangeToRowIndex(relativeRangeId(-lowerOffset)),
+                  rangeToRowIndex(relativeRangeId(upperOffset))
+              )
+          );
+
+          currentRowIndex = rangeToRowIndex(currentRangeIndex + 1);
+          currentRangeIndex++;
+          return r;
+        }
+
+        private int rangeToRowIndex(int rangeId)
+        {
+          return rangeToRowId[rangeId];
+        }
+
+        private int relativeRangeId(int rangeOffset)
+        {
+          int rangeId = currentRangeIndex + rangeOffset;
+          if (rangeId < 0) {
+            return 0;
+          }
+          if (rangeId >= numRanges) {
+            return numRanges;
+          }
+          return rangeId;
+        }
+      };
+    }
+  }
+
+
+  /**
+   * Basic [a,b) interval; left inclusive/right exclusive.
+   */
+  static class Interval implements Iterable<Integer>
+  {
+    int a;
+    int b;
+
+    public static Interval of(int u, int v)
+    {
+      return new Interval(u, v);
+    }
+
+    public Interval(int u, int v)
+    {
+      this.a = u;
+      this.b = v;
+    }
+
+    @Override
+    public Iterator<Integer> iterator()
+    {
+      return new Iterator<Integer>()
+      {
+        int current = a;
+
+        @Override
+        public Integer next()
+        {
+          if (!hasNext()) {
+            throw new IllegalStateException();
+          }
+          return current++;
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+          return current < b;
+        }
+      };
+
+    }
+  }
+
+  /**
+   * Represents a range between U (inclusive) and V (exclusive).
+   */
+  static class Range
+  {
+    Interval outputRows;
+    Interval inputRows;
+
+    public Range(Interval outputRows, Interval inputRows)
+    {
+      this.outputRows = outputRows;
+      this.inputRows = inputRows;
+    }
+  }
+
+  static class AggCell
+  {
+    private AggregatorFactory[] aggFactories;
+    Interval currentRows = new Interval(0, 0);
+    private final AtomicInteger rowIdProvider;
+    private final ColumnSelectorFactory columnSelectorFactory;
+
+    private final Aggregator[] aggregators;
+
+
+    AggCell(AppendableRowsAndColumns rac, AggregatorFactory[] aggFactories)
```

Comment:
Instead of this depending on an `AppendableRowsAndColumns` I'd suggest that you make it depend on a `ColumnSelectorFactoryMaker`.  On top of that, if it was me, I'd write it as a static creator method and have this constructor take the `AtomicInteger, ColumnSelectorFactory, AggregatorFactory[]` tuple.